### PR TITLE
Crop image effect

### PR DESCRIPTION
**Pull request by slashrsm**
https://www.drupal.org/node/2376691

### CODE REVIEW

File: crop.module [hunk 1]
```diff
@@ -6,3 +6,25 @@
...
  */
+
+
+/**
+ * Implements hook_theme().
+ */
+function crop_theme() {
+  return [
+    'crop_crop_summary' => array(
+      'variables' => array('data' => NULL, 'effect' => array()),
```
Comment (marcingy) [inline crop.module]: Should data be an empty array?

File: crop.module [hunk 2]
```diff
@@ -6,3 +6,25 @@
...
+  ];
+}
+
+/**
+ * Prepares variables for crop_crop summary template.
+ *
+ * Default template: crop-crop-summary.twig.html
+ */
+function template_preprocess_crop_crop_summary(&$variables) {
+  $type = \Drupal::entityManager()->getStorage('crop_type')->load($variables['data']['crop_type']);
```
Comment (marcingy) [inline crop.module]: $variables['data']['crop_type'] could notice given the default should you instalise in structure?

File: src/Entity/Crop.php [hunk 3]
```diff
@@ -60,6 +60,36 @@ class Crop extends ContentEntityBase implements CropInterface {
...
+      'y' => $this->y->value,
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function anchor() {
+    return [
+      'x' => $this->x->value - $this->width->value / 2,
```
Comment (marcingy) [inline src/Entity/Crop.php]: Maybe add () to make this more explict as to opperations

File: src/Plugin/ImageEffect/CropEffect.php [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function applyEffect(ImageInterface $image) {
+    if (empty($this->configuration['crop_type']) || !$this->typeStorage->load($this->configuration['crop_type'])) {
+      $this->logger->error('Manual image crop failed due to misconfigured crop type on %path.', ['%path' => $image->getSource()]);
+      return FALSE;
+    }
+
+    if (($crop = $this->getCrop($image))) {
```
Comment (marcingy) [inline src/Plugin/ImageEffect/CropEffect.php]: Extra level of ()

File: src/Plugin/ImageEffect/CropEffect.php [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+      $anchor = $crop->anchor();
+      $size = $crop->size();
+
+      if (!$image->crop($anchor['x'], $anchor['y'], $size['width'], $size['height'])) {
+        $this->logger->error('Manual image crop failed using the %toolkit toolkit on %path (%mimetype, %dimensions)', [
+            '%toolkit' => $image->getToolkitId(),
+            '%path' => $image->getSource(),
+            '%mimetype' => $image->getMimeType(),
+            '%dimensions' => $image->getWidth() . 'x' . $image->getHeight()
```
Comment (marcingy) [inline src/Plugin/ImageEffect/CropEffect.php]: Why not have %width and %height and move the x into the log entry - I am thinking that might make translation easier? If it does then this is ok

File: src/Plugin/ImageEffect/CropEffect.php [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    foreach ($this->typeStorage->loadMultiple() as $type) {
+      $options[$type->id()] = $type->label();
+    }
+
+    $form['crop_type'] = [
+      '#type' => 'select',
+      '#title' => t('Crop type'),
+      '#default_value' => $this->configuration['crop_type'],
+      '#options' => $options,
+      '#description' => t("Crop type to be used for the image style."),
```
Comment (marcingy) [inline src/Plugin/ImageEffect/CropEffect.php]: Lets use ' vs " - nitpik but I assume ' should be used except where the string includes ' itself

File: src/Plugin/ImageEffect/CropEffect.php [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+   *   Crop entity or FALSE if crop doesn't exist.
+   */
+  protected function getCrop(ImageInterface $image) {
+    if (!isset($this->crop)) {
+      $this->crop = FALSE;
+
+      $ids = $this->query
+        ->condition('uri', $image->getSource())
+        ->condition('type', $this->configuration['crop_type'])
+        ->sort('cid')
```
Comment (marcingy) [inline src/Plugin/ImageEffect/CropEffect.php:178]: Is the sort needed? Surely the type is a machine name and hence unique?

File: src/Plugin/ImageEffect/CropEffect.php [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+   * @param ImageInterface $image
+   *   Image object.
+   * @return \Drupal\Core\Entity\EntityInterface|\Drupal\crop\CropInterface|FALSE
+   *   Crop entity or FALSE if crop doesn't exist.
+   */
+  protected function getCrop(ImageInterface $image) {
+    if (!isset($this->crop)) {
+      $this->crop = FALSE;
+
+      $ids = $this->query
```
Comment (marcingy) [inline src/Plugin/ImageEffect/CropEffect.php]: $id is more what you get back?

File: src/Tests/CropEffectTest.php [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    $this->assertTrue(file_exists($derivative_uri), 'Image derivative file exists on the filesystem.');
+
+    // Test if cropped version looks like expected. Basically loop pixels in derivative
+    // image and check if they look the same as pixels in corresponding region
+    // on original image.
+    $original_image = imagecreatefrompng($file->getFileUri());
+    $derivative_image = imagecreatefrompng($derivative_uri);
+    $orig_start = $crop->anchor();
+    $matches = TRUE;
+    for ($x = 0; $x < $values['width']; $x++) {
```
Comment (marcingy) [inline src/Tests/CropEffectTest.php:68]: $x why not $width and for $y $height

File: src/Tests/CropUnitTestBase.php [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    $entity_manager->onEntityTypeCreate($entity_manager->getDefinition('user'));
+    $entity_manager->onEntityTypeCreate($entity_manager->getDefinition('image_style'));
+    $entity_manager->onEntityTypeCreate($entity_manager->getDefinition('crop'));
+
+    // Create test image style
+    $this->testStyle = $this->imageStyleStorage->create([
+      'name' => 'test',
+      'label' => 'Test image style',
+      'effects' => [
+        '69ad20fb-13b0-4e84-acaf-ddd91e7e15ac' => [
```
Comment (marcingy) [inline src/Tests/CropUnitTestBase.php]: Should we be providing UUIDS? Both here and below

File: src/Plugin/ImageEffect/CropEffect.php (hunk 7 above)
Comment (slashrsm) [inline src/Plugin/ImageEffect/CropEffect.php:178]: This is not entity type. It is bundle and is not unique. Added sort to assure deterministic behaviour.

File: src/Tests/CropEffectTest.php (hunk 9 above)
Comment (slashrsm) [inline src/Tests/CropEffectTest.php:68]: width, height are size information. x,y are coordinates which is what we have in this specific situation.

**Pull request merged**